components/BlurhashCanvas.vue

```diff
@@ -37,15 +38,21 @@ export default Vue.extend({
   },
   methods: {
     draw() {
-      const pixels = decode(this.hash, this.width, this.height);
-      if (pixels) {
-        const ctx = (this.$refs.canvas as HTMLCanvasElement).getContext('2d');
-        const imageData = ctx?.createImageData(this.width, this.height);
-        if (imageData) {
+      const bhash_data = {
+        hash: this.hash,
+        width: this.width,
+        height: this.height
+      };
+      worker.postMessage(bhash_data);
+      const ctx = (this.$refs.canvas as HTMLCanvasElement).getContext('2d');
+      const imageData = ctx?.createImageData(this.width, this.height);
+      worker.onmessage = ({ data }) => {
```

Comment:
I think a better way to do this would be to set a watcher on the hash and size, which would send the message to the worker, then define the onmessage in "created" (since it's an event, it seems to be what you're supposed to do) and have it call upon the `draw` method with the pixel array.

Reply:
I've been all day and afternoon trying to do this and I couldn't. What I tried:

* Define a ``pixels`` data() prop => **Data prop seems to be shared across BlurhashCanvas instances, so consequent instances called after one instance would receive the results from the worker, applying the pixels of the first element in the row to the last one, for instance.
* Use a watcher => I either did it wrong or Vue seems to "fire and forget" about the canvas, so same thing happened as in the first attempt: we were placing pixels in the wrong cards.
* Multiple combinations I can't remember now.

Everything led me to the conclusion that ``this.$refs.canvas as HTMLCanvasElement`` points to the DOM node that it's currently being created, so the event handler will always lag behind, applying it to the incorrect node, leaving some cards without blurhash. It's strange though, because the ``onmessage`` function never reported having anything ``undefined`` and I tracked the process with logging everywhere and everything seemed to be right (except from the canvas that draw the pixels!).

**How I solved it?

It's ugly, but the only solution I found was to add the canvas node to a dict where the hash is the key. In the ``onmessage`` event, we receive the original hash and the pixels from the worker. With this, I get the canvas node and draw the pixels. Once it's done, I remove the element in the array.

I also complemented this by storing the decoded pixels in another array. Not sure if this is being caused by Virtual Scroller switching DOMs, but every time you scroll and see items that were not previously in the viewport, you'll see that they're generating blurhashes, which is unefficient. I didn't cover the fact that the array has no limits so it can eat up a lot of memory (wanted first to know your opinions in how to deal with this and how to solve this whole problem) but it seems to be very effective: Take in mind that the array where we "cache" the blurhashes is defined globally for the ``BlurhashCanvas`` component, like the worker, so every blurhash of every image will be hold on memory as soon as the page is reloaded. The memory footprint seems to be really low: I scrolled through my entire movie collection (~500 items), TV Shows (~100 items) and music (+1200 items) multiple times and the highest memory usage recorded by Chrome debugger was between **200-450 MB**, in a cycle (I guess the spikes were caused by Virtual Scroller recycling the DOM).

Reply:
@MrTimscampi See [here](https://github.com/jellyfin/jellyfin-vue/pull/213/files#diff-fdffe0dcc16277333627cbefb4982d475d27eb2f237f92efb9376dda66d5a8ebL31-R57)